### PR TITLE
[featuregate] Deprecate NewFlag in favor of RegisterFlags

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

Deprecate `featuregate.NewFlag` in favor of `Registry.RegisterFlags`, which registers flags into a provided flagset.

The intent is to unblock future development of new features such as #7991 without incurring into deprecations or breaking changes to implement these.
This will allow us to do a 1.0 release.

**Link to tracking Issue:** Relates to #8220

### CODE REVIEW

**Review by djaglowski — APPROVED** [review]

**Review by TylerHelmuth — APPROVED** [review]

**Review by MovieStoreGuy — APPROVED** [review]

File: featuregate/flag.go [hunk 1]
```diff
@@ -10,11 +10,25 @@ import (
 	"go.uber.org/multierr"
 )
 
+// Flags related to feature gates.
+type Flags struct {
```
Comment (bogdandrutu) [inline featuregate/flag.go:14]: Question, why just having Options in the `NewFlag` is not enough to extend capabilities?

Comment (mx-psi) [reply featuregate/flag.go:14]: >  why just having Options in the NewFlag is not enough to extend capabilities?

How would you add the new flag from #7991 in this way? One would have to define the strict flag before defining the usual flag, I am not sure if that is even possible.

I am also not sure what the benefit of `Options` over this approach is. It is still [a breaking change](https://go.dev/play/p/LYJ7VL_kXwP).

Comment (bogdandrutu) [reply featuregate/flag.go:14]: I am not arguing about breaking change, but about the fact that now this API is more complicated for normal users. Returning a struct that then I need decide which flag to install, etc.

If we want to allow use to construct multiple flags, then pass a FlagSet to the func and do that. Also still we should add Options to allow extensions in the future.

**Comment by atoulme** [comment]
I think I need to simplify things. I will drop this requirement to run this with a CLI flag since it's so impractical and work my way with a programmatic change so I can test strict mode in a test. We can stop this, we're trying to make it work but it's just impractical and I have exhausted enough good will with this attempt.

**Comment by bogdandrutu** [comment]
@atoulme I understand your frustration, but please don't close others PRs :)

**Comment by atoulme** [comment]
@bogdandrutu Oh I meant to help close this because @mx-psi only opened it to make my own PR whole as we're going for 1.0. No harm done. Just trying to make it explicit to @mx-psi his help is deeply appreciated, but I need to relieve him from his duty of pushing this work in ahead of the 1.0 release of featuregate. I'll go close my original PR as well.

File: featuregate/flag.go (hunk 1 above)
Comment (bogdandrutu) [reply featuregate/flag.go:14]: I would do something like:

```golang
// NewFlags returns Flags that control building the global Registry.
func RegisterFlag(flag.FlagSet fs, reg *Registry, opts ...Option) Flags {
	...
        // One option can be to enable strict mode/to add a flag for that.
       // Though, the problem with flag is that it has to be defined first.
}
```

**Comment by bogdandrutu** [comment]
@mx-psi @atoulme I still believe the current `NewFlag` is not very extensible, so having a way to extend that may help in the future, see my comments.

**Comment by atoulme** [comment]
Understood! I'll leave it be then. But I can decorrelate the work on strict mode from flags, for now at least.

File: featuregate/flag.go (hunk 1 above)
Comment (mx-psi) [reply featuregate/flag.go]: PTAL at 1fb858c67, is that what you had in mind? This works for me

**Review by bogdandrutu — APPROVED** [review]

File: featuregate/flag.go [hunk 2]
```diff
@@ -10,11 +10,31 @@ import (
...
 
+// FlagsOption is an option for RegisterFlags.
+type FlagsOption interface {
+	private()
+}
+
+// RegisterFlags that directly applies feature gate statuses to a Registry.
+func RegisterFlags(flagSet *flag.FlagSet, reg *Registry, opts ...FlagsOption) {
+	flagSet.Var(newFeatureGateValue(reg), featureGatesFlag, featureGatesFlagDescription)
+}
```
Comment (bogdandrutu) [inline featuregate/flag.go:36]: @mx-psi sorry for this :).

Do you think we can simplify this API and make this a func on the Registry struct? Makes things simpler. 

Also separate concern: FlagsOption -> RegisterFlagsOption

Comment (mx-psi) [reply featuregate/flag.go]: Tried that out on f20fbf9e80216aa468b541614b1cdc494856b36e, looks nice :)

**Comment by mx-psi** [comment]
@atoulme @TylerHelmuth @djaglowski PR changed a bit after back and forth with Bogdan, mind taking another look?

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**